language/stdlib/modules/transaction_fee_collection.mvir

```diff
@@ -0,0 +1,86 @@
+    current_height: u64,
+  }
+
+  // Initialize the fee pot, and set the current to the current height of
+  // the blockchain. This should be 0 since this is only called in genesis.
+  public initialize() {
+    assert(get_txn_sender() == 0xA550C18, 0);
+    move_to_sender<CollectedFees>(CollectedFees {
+        fees: LibraCoin.zero(),
+        current_height: Block.get_current_height(),
+      });
+    return;
+  }
+
+  // Pay a transaction fee into the the transaction fee pot.
+  // TODO/FUTURE: We may want a way to guard the calls to this function so
+  //   that not just anyone in any context can pay into the fee pot.  We need
+  //   to look into the economic effects that having an open
+  //   "pay_transaction_fee" may have (e.g. by over-incentivizing
+  //   processing). This will guide whether or not this function needs to be
+  //   privatized or not.
+  //
+  // One way to restrict the access to calling this function is to
+  // inline this into the LibraAccount, but this has a code smell since
+  // this logic shouldn't really be coupled with every libra account just
+  // so that a single resource can be published under a single account.
+  public pay_transaction_fee(fee_to_deposit: LibraCoin.T) acquires CollectedFees {
+    let collected_fees: &mut Self.CollectedFees;
+    collected_fees = borrow_global_mut<CollectedFees>(0xA550C18);
+    LibraCoin.deposit(&mut move(collected_fees).fees, move(fee_to_deposit));
+    return;
+  }
+
+  // Extract the transaction fees that have been collected. Anyone can call
+  // this function, but the only caller permitted to extract the fees is
+  // the association account. All other calls will lead to an abort.
+  public extract_transaction_fees(): LibraCoin.T acquires CollectedFees {
+      let current_height: u64;
+      let lib_coin: LibraCoin.T;
+      let amount_collected: u64;
+      let collected_fees: &mut Self.CollectedFees;
+      let collected_fees_current_height: &mut u64;
+      let lib_coin_to_distribute: LibraCoin.T;
+
+      assert(get_txn_sender() == 0xA550C18, 1);
+      current_height = Block.get_current_height();
+      collected_fees = borrow_global_mut<CollectedFees>(0xA550C18);
+      collected_fees_current_height = &mut copy(collected_fees).current_height;
+      amount_collected = LibraCoin.value(&copy(collected_fees).fees);
+
+      // If the collected_fees has already been paid out for this _block_, then don't do anything.
+      if (copy(current_height) < *copy(collected_fees_current_height)) {
+          return LibraCoin.zero();
+      }
+
+      // Otherwise, update the height, and then distribute the fees.
+      *copy(collected_fees_current_height) = move(current_height);
+
+      lib_coin_to_distribute = LibraCoin.withdraw(
+          &mut move(collected_fees).fees,
+          copy(amount_collected)
+      );
+
+      return move(lib_coin_to_distribute);
+  }
+
+  public fees_collected_balance_FOR_TESTING_ONLY(): u64 acquires CollectedFees {
```

Comment:
uh... maybe we should have some different system here for testing?

Reply:
I don't like this either. However, I can't come up with another easy/simple way of testing it without this, or relaxing some of the design (which is even worse IMO). 

Also, having this ability, not even as just a testing function might not be bad to have in the final module definition since this would e.g. allow validators to query how much they can currently expect in transaction fees possibly. But that's a larger design decision I'd like to discuss separately from this if possible.